File: unity/Assets/Scripts/Combat/Wind.cs
// <copyright file="Wind.cs" company="board.fun game">
//     Trafalgar - Age of Sail. Procedural board.fun RTS.
// </copyright>

namespace Trafalgar.Combat
{
    using Trafalgar.Core;
    using UnityEngine;

    /// <summary>
    /// The global wind. Age-of-sail ships cannot sail directly into the wind, so wind direction
    /// is the central tactical constraint. The wind slowly veers over the course of a battle.
    /// </summary>
    public class Wind
    {
        private float m_ShiftTimer;
        private float m_TargetFromDeg;

        /// <summary>
        /// Initialises a new instance of the <see cref="Wind"/> class.
        /// </summary>
        /// <param name="initialFromDeg">Initial direction the wind blows *from*, in compass degrees.</param>
        public Wind(float initialFromDeg)
        {
            FromDegrees = Nav.Normalize360(initialFromDeg);
            m_TargetFromDeg = FromDegrees;
            m_ShiftTimer = GameConfig.WindShiftInterval;
        }

        /// <summary>Gets the direction the wind blows *from*, in compass degrees (the "weather gauge").</summary>
        public float FromDegrees { get; private set; }

        /// <summary>Gets a unit vector pointing in the direction the wind is blowing *toward* (downwind).</summary>
        public Vector3 BlowingToward => Nav.HeadingToVector(Nav.Normalize360(FromDegrees + 180f));

        /// <summary>Gets a unit vector pointing toward the source of the wind (upwind).</summary>
        public Vector3 Source => Nav.HeadingToVector(FromDegrees);

        /// <summary>
        /// Advances the wind simulation, occasionally veering the direction.
        /// </summary>
        /// <param name="dt">Delta time in seconds.</param>
        public void Tick(float dt)
        {
            m_ShiftTimer -= dt;
            if (m_ShiftTimer <= 0f)
            {
                m_ShiftTimer = GameConfig.WindShiftInterval;
                float shift = Random.Range(-GameConfig.WindShiftMagnitude, GameConfig.WindShiftMagnitude);
                m_TargetFromDeg = Nav.Normalize360(m_TargetFromDeg + shift);
            }

            // Ease toward the target direction so shifts feel gradual.
            FromDegrees = Nav.MoveTowardsAngle(FromDegrees, m_TargetFromDeg, 3f * dt);
        }

        /// <summary>
        /// Computes the speed multiplier for a ship on a given heading (its "point of sail").
        /// </summary>
        /// <param name="headingDeg">The ship's heading in compass degrees.</param>
        /// <returns>A multiplier in [<see cref="GameConfig.InIronsFactor"/>, 1].</returns>
        public float PointOfSailFactor(float headingDeg)
        {
            // Angle between where we are pointing and the wind's source.
            // 0   => sailing straight into the wind (in irons)
            // 180 => running dead downwind
            float offWind = Nav.AngleDifference(headingDeg, FromDegrees);
            return PointOfSailFactor(offWind, out _);
        }

        /// <summary>
        /// Computes the speed multiplier and classifies the point of sail.
        /// </summary>
        /// <param name="offWindAngle">Angle between heading and the wind source, in [0, 180].</param>
        /// <param name="pointOfSail">Outputs a friendly classification.</param>
        /// <returns>A speed multiplier in [<see cref="GameConfig.InIronsFactor"/>, 1].</returns>
        public static float PointOfSailFactor(float offWindAngle, out string pointOfSail)
        {
            offWindAngle = Mathf.Abs(offWindAngle);

            if (offWindAngle < GameConfig.NoGoAngle)
            {
                pointOfSail = "In Irons";

                // Ramp from the floor at dead-into-wind up toward close-hauled.
                float t = Mathf.InverseLerp(0f, GameConfig.NoGoAngle, offWindAngle);
                return Mathf.Lerp(GameConfig.InIronsFactor, 0.45f, t);
            }

            if (offWindAngle < 75f)
            {
                pointOfSail = "Close-Hauled";
                return Mathf.Lerp(0.45f, 0.85f, Mathf.InverseLerp(GameConfig.NoGoAngle, 75f, offWindAngle));
            }

            if (offWindAngle < 115f)
            {
                pointOfSail = "Beam Reach";
                return Mathf.Lerp(0.85f, 1.0f, Mathf.InverseLerp(75f, 100f, offWindAngle));
            }

            if (offWindAngle < 150f)
            {
                pointOfSail = "Broad Reach";
                return Mathf.Lerp(1.0f, 0.9f, Mathf.InverseLerp(115f, 150f, offWindAngle));
            }

            pointOfSail = "Running";
            return Mathf.Lerp(0.9f, 0.78f, Mathf.InverseLerp(150f, 180f, offWindAngle));
        }
    }
}

File: web/src/combat/wind.ts
// The global wind — a port of Unity `Combat/Wind.cs`. Age-of-sail ships cannot
// sail directly into the wind, so wind direction is the central tactical
// constraint. The wind slowly veers over the course of a battle.

import * as Config from "../core/config";
import { headingToVector, normalize360, angleDifference, moveTowardsAngle } from "../core/nav";
import { inverseLerp, lerp } from "../core/mathf";
import { rangeFloat } from "../core/rng";
import type { Vec2 } from "../core/vec";

export interface PointOfSailResult {
  factor: number;
  pointOfSail: string;
}

export class Wind {
  /** Direction the wind blows *from*, in compass degrees (the "weather gauge"). */
  fromDegrees: number;

  private shiftTimer: number;
  private targetFromDeg: number;

  constructor(initialFromDeg: number) {
    this.fromDegrees = normalize360(initialFromDeg);
    this.targetFromDeg = this.fromDegrees;
    this.shiftTimer = Config.WindShiftInterval;
  }

  /** Unit vector pointing the way the wind is blowing *toward* (downwind). */
  get blowingToward(): Vec2 {
    return headingToVector(normalize360(this.fromDegrees + 180));
  }

  /** Unit vector pointing toward the source of the wind (upwind). */
  get source(): Vec2 {
    return headingToVector(this.fromDegrees);
  }

  tick(dt: number): void {
    this.shiftTimer -= dt;
    if (this.shiftTimer <= 0) {
      this.shiftTimer = Config.WindShiftInterval;
      const shift = rangeFloat(-Config.WindShiftMagnitude, Config.WindShiftMagnitude);
      this.targetFromDeg = normalize360(this.targetFromDeg + shift);
    }

    this.fromDegrees = moveTowardsAngle(this.fromDegrees, this.targetFromDeg, 3 * dt);
  }

  /** Speed multiplier for a ship on a given heading (its "point of sail"). */
  pointOfSailFactorFor(headingDeg: number): number {
    const offWind = angleDifference(headingDeg, this.fromDegrees);
    return pointOfSailFactor(offWind).factor;
  }
}

// Point-of-sail status colours — shared by the on-ship point-of-sail dot AND the
// course-vector lines so the two always read the same. A SIX-step slow→fast ramp,
// one colour per point of sail (see `pointOfSailColorIndex` / `pointOfSailFactor`):
//   0 In Irons      → red       (no-go, slowest)
//   1 Close-Hauled  → orange
//   2 Close Reach   → amber/yellow
//   3 Beam Reach    → yellow-green
//   4 Broad Reach   → green      (fastest)
//   5 Running       → teal/cyan  (fast, slightly off the broad-reach peak)
export const POINT_OF_SAIL_COLORS = [
  0xff4d4d, 0xff8c33, 0xffd24a, 0x9ed94d, 0x4dd06a, 0x33c4c4,
] as const;

/** Maps a point-of-sail label to its colour-ramp index (0..5). Every label
 *  `pointOfSailFactor` can return is covered; unknown labels fall back to a
 *  mid/fast green so there's never an out-of-range lookup. */
export function pointOfSailColorIndex(label: string): number {
  switch (label) {
    case "In Irons":
      return 0;
    case "Close-Hauled":
      return 1;
    case "Close Reach":
      return 2;
    case "Beam Reach":
      return 3;
    case "Broad Reach":
      return 4;
    case "Running":
      return 5;
    default:
      return 4;
  }
}

/** The point-of-sail status colour for sailing `headingDeg` in the given wind —
 *  used to tint the course preview by the expected speed of that heading. */
export function pointOfSailColor(headingDeg: number, wind: Wind): number {
  const off = angleDifference(headingDeg, wind.fromDegrees);
  return POINT_OF_SAIL_COLORS[pointOfSailColorIndex(pointOfSailFactor(off).pointOfSail)];
}

/**
 * Computes the speed multiplier and classifies the point of sail across SIX
 * classic bands (angle off the true wind, 0° = dead into wind → 180° = dead
 * downwind). Speeds rise from the no-go zone to a broad-reach peak, then ease
 * back a touch dead downwind (square-rig blanketing). Factors interpolate
 * smoothly and are continuous at every band boundary:
 *
 *   In Irons     0°–NoGoAngle(42°)   InIronsFactor → 0.30   (slowest, no-go)
 *   Close-Hauled 42°–60°             0.30 → 0.50
 *   Close Reach  60°–80°             0.50 → 0.70
 *   Beam Reach   80°–100°            0.70 → 0.88 → 0.90    (bulged at 90°)
 *   Broad Reach  100°–150°           0.90 → 1.00            (fastest)
 *   Running      150°–180°           1.00 → 0.85
 */
export function pointOfSailFactor(offWindAngle: number): PointOfSailResult {
  offWindAngle = Math.abs(offWindAngle);

  if (offWindAngle < Config.NoGoAngle) {
    return {
      pointOfSail: "In Irons",
      factor: lerp(Config.InIronsFactor, 0.3, inverseLerp(0, Config.NoGoAngle, offWindAngle)),
    };
  }

  if (offWindAngle < 60) {
    return {
      pointOfSail: "Close-Hauled",
      factor: lerp(0.3, 0.5, inverseLerp(Config.NoGoAngle, 60, offWindAngle)),
    };
  }

  if (offWindAngle < 80) {
    return {
      pointOfSail: "Close Reach",
      factor: lerp(0.5, 0.7, inverseLerp(60, 80, offWindAngle)),
    };
  }

  if (offWindAngle < 100) {
    // Bulged at the true beam (90°): ships hold more speed perpendicular to the
    // wind. Endpoints (0.70 at 80°, 0.90 at 100°) match the neighbouring bands,
    // so the curve stays continuous.
    const factor =
      offWindAngle < 90
        ? lerp(0.7, 0.88, inverseLerp(80, 90, offWindAngle))
        : lerp(0.88, 0.9, inverseLerp(90, 100, offWindAngle));
    return { pointOfSail: "Beam Reach", factor };
  }

  if (offWindAngle < 150) {
    return {
      pointOfSail: "Broad Reach",
      factor: lerp(0.9, 1.0, inverseLerp(100, 150, offWindAngle)),
    };
  }

  return {
    pointOfSail: "Running",
    factor: lerp(1.0, 0.85, inverseLerp(150, 180, offWindAngle)),
  };
}
